sdk/metric/batcher/defaultkeys/defaultkeys.go

```diff
@@ -0,0 +1,145 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package defaultkeys // import "go.opentelemetry.io/otel/sdk/metric/batcher/defaultkeys"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/api/core"
+	export "go.opentelemetry.io/otel/sdk/export/metric"
+)
+
+type (
+	Batcher struct {
+		selector export.AggregationSelector
+		lencoder export.LabelEncoder
+		stateful bool
+		dki      dkiMap
+		agg      aggMap
+	}
+
+	aggEntry struct {
+		descriptor *export.Descriptor
+		labels     export.Labels
+		aggregator export.Aggregator
+	}
+
+	dkiMap map[*export.Descriptor]map[core.Key]int
+	aggMap map[string]aggEntry
+
+	producer struct {
+		aggMap   aggMap
+		lencoder export.LabelEncoder
+	}
+)
+
+var _ export.Batcher = &Batcher{}
+var _ export.Producer = &producer{}
+
+func New(selector export.AggregationSelector, lencoder export.LabelEncoder, stateful bool) *Batcher {
+	return &Batcher{
+		selector: selector,
+		lencoder: lencoder,
+		dki:      dkiMap{},
+		agg:      aggMap{},
+		stateful: stateful,
+	}
+}
+
+func (b *Batcher) AggregatorFor(descriptor *export.Descriptor) export.Aggregator {
+	return b.selector.AggregatorFor(descriptor)
+}
+
+func (b *Batcher) Process(_ context.Context, desc *export.Descriptor, labels export.Labels, agg export.Aggregator) error {
+	keys := desc.Keys()
+
+	// Cache the mapping from Descriptor->Key->Index
+	ki, ok := b.dki[desc]
+	if !ok {
+		ki = map[core.Key]int{}
+		b.dki[desc] = ki
+
+		for i, k := range keys {
+			ki[k] = i
+		}
+	}
+
+	// Compute the value list.  Note: Unspecified values become
+	// empty strings.  TODO: pin this down, we have no appropriate
+	// Value constructor.
+	canon := make([]core.KeyValue, len(keys))
+
+	for i, key := range keys {
+		canon[i] = key.String("")
+	}
+
+	// Note also the possibility to speed this computation of
+	// "encoded" via "canon" in the form of a (Descriptor,
+	// LabelSet)->(Labels, Encoded) cache.
+	for _, kv := range labels.Ordered() {
+		pos, ok := ki[kv.Key]
+		if !ok {
+			continue
+		}
+		canon[pos].Value = kv.Value
+	}
+
+	// Compute an encoded lookup key.
+	encoded := b.lencoder.EncodeLabels(canon)
+
+	// Reduce dimensionality.
+	rag, ok := b.agg[encoded]
+	if !ok {
+		// If this Batcher is stateful, create a copy of the
+		// Aggregator for long-term storage.  Otherwise the
+		// Meter implementation will checkpoint the aggregator
+		// again, overwriting the long-lived state.
+		if b.stateful {
+			tmp := agg
+			agg = b.AggregatorFor(desc)
+			if err := agg.Merge(tmp, desc); err != nil {
+				return err
+			}
+		}
+		b.agg[encoded] = aggEntry{
+			descriptor: desc,
+			labels:     export.NewLabels(canon, encoded, b.lencoder),
+			aggregator: agg,
+		}
+		return nil
+	}
+	return rag.aggregator.Merge(agg, desc)
+}
+
+func (b *Batcher) ReadCheckpoint() export.Producer {
```

Comment:
this is confusing that a `Read` operation would change my state by clearing `agg`; should be explicit that if we are not stateful, then we reset our values here.